### PR TITLE
[SPARK-39927][BUILD] Upgrade to Avro 1.11.1

### PR DESCRIPTION
**Pull request by iemejia**
### What changes were proposed in this pull request?
Update the Avro version to 1.11.1

### Why are the changes needed?
To stay up to date with upstream

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Unit tests

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Seems like the test failure looks related:

```
AvroV1Suite.support user provided avro schema for writing non-nullable enum type
java.lang.AssertionError: assertion failed: Exception tree doesn't contain the expected exception of type org.apache.avro.AvroTypeException with message: Not an enum: null
org.apache.avro.AvroTypeException: value null is not a SuitEnumType
	at org.apache.avro.generic.GenericDatumWriter.writeEnum(GenericDatumWriter.java:269)
	at org.apache.avro.specific.SpecificDatumWriter.writeEnum(SpecificDatumWriter.java:66)
	at org.apache.avro.generic.GenericDatumWriter.writeWithoutConversion(GenericDatumWriter.java:148)
	at org.apache.avro.generic.GenericDatumWriter.write(GenericDatumWriter.java:95)
	at org.apache.avro.reflect.ReflectDatumWriter.write(ReflectDatumWriter.java:158)
	at org.apache.avro.generic.GenericDatumWriter.writeField(GenericDatumWriter.java:245)
	at org.apache.avro.specific.SpecificDatumWriter.writeField(SpecificDatumWriter.java:117)
	at org.apache.avro.reflect.ReflectDatumWriter.writeField(ReflectDatumWriter.java:184)
	at org.apache.avro.generic.GenericDatumWriter.writeRecord(GenericDatumWriter.java:234)
	at org.apache.avro.specific.SpecificDatumWriter.writeRecord(SpecificDatumWriter.java:92)
	at org.apache.avro.generic.GenericDatumWriter.writeWithoutConversion(GenericDatumWriter.java:145)
	at org.apache.avro.generic.GenericDatumWriter.write(GenericDatumWriter.java:95)
	at org.apache.avro.reflect.ReflectDatumWriter.write(ReflectDatumWriter.java:158)
	at org.apache.avro.generic.GenericDatumWriter.write(GenericDatumWriter.java:82)
	at org.apache.avro.file.DataFileWriter.append(DataFileWriter.java:314)
	at org.apache.spark.sql.avro.SparkAvroKeyRecordWriter.write(SparkAvroKeyOutputFormat.java:86)
	at org.apache.spark.sql.avro.SparkAvroKeyRecordWriter.write(SparkAvroKeyOutputFormat.java:63)
	at org.apache.spark.sql.avro.AvroOutputWriter.write(AvroOutputWriter.scala:86)
	at org.apache.spark.sql.execution.datasources.SingleDirectoryDataWriter.write(FileFormatDataWriter.scala:175)
	at org.apache.spark.sql.execution.datasources.FileFormatDataWriter.writeWithMetrics(FileFormatDataWriter.scala:85)
	at org.apache.spark.sql.execution.datasources.FileFormatDataWriter.writeWithIterator(FileFormatDataWriter.scala:92)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$.$anonfun$executeTask$1(FileFormatWriter.scala:323)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1524)
```

**Comment by xkrogen** [comment]
Is this an official release? The documentation links you updated ([such as this one](https://avro.apache.org/docs/1.11.1/spec.html#schema_record)) give a 404 error and I don't see Avro 1.11.1 listed on the [Avro Releases page](https://avro.apache.org/releases.html).

**Review by dongjoon-hyun — COMMENTED** [review]
- Thank you for making a PR, @iemejia .
- Unfortunately, as mentioned by Hyukjin, for the failures, it looks like that.
  - It seems that Avro error message is changed. Could you adjust the error message assertion part in Spark test case according to Avro change, please, @iemejia ?
- To @xkrogen . Yes, right, but @iemejia published it. Usually, it takes some time to sync.
  - https://repo1.maven.org/maven2/org/apache/avro/avro/1.11.1/

**Comment by iemejia** [comment]
@xkrogen It is indeed. The announcement has not gone out yet but the binaries are already available.
https://lists.apache.org/thread/8tk92x804owmjbvtj57xxzwhxn1qh3ho

**Comment by dongjoon-hyun** [comment]
Gentle ping, @iemejia .

**Comment by martin-g** [comment]
> I don't see Avro 1.11.1 listed on the [Avro Releases page](https://avro.apache.org/releases.html).

Also, we work on replacing the Avro website with a new one. Probably it will be done later this week!

**Comment by LuciferYang** [comment]
seems `support user provided non-nullable avro schema for nullable catalyst schema without any null record *** FAILED *** (181 milliseconds)` still failed

**Comment by dongjoon-hyun** [comment]
Yes, it does. @LuciferYang .

To @iemejia , as I mentioned [here](https://github.com/apache/spark/pull/37352#pullrequestreview-1057658777), please take a look at the test case. Otherwise, Avro 1.11.1 cannot pass the CIs.

File: connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala [hunk 1]
```diff
@@ -875,7 +875,7 @@ abstract class AvroSuite
         dfWithNull.write.format("avro")
           .option("avroSchema", avroSchema).save(s"$tempDir/${UUID.randomUUID()}")
       }
-      assertExceptionMsg[AvroTypeException](e1, "Not an enum: null")
+      assertExceptionMsg[AvroTypeException](e1, "value null is not a SuitEnumType")
```
Comment (dongjoon-hyun) [inline connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala:878]: Thank you for the fix.

File: connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala [hunk 2]
```diff
@@ -1075,8 +1075,7 @@ abstract class AvroSuite
           .save(s"$tempDir/${UUID.randomUUID()}")
       }.getCause.getMessage
       assert(message.contains("Caused by: java.lang.NullPointerException: "))
-      assert(message.contains(
-        "null of string in string in field Name of test_schema in test_schema"))
+      assert(message.contains("null in string in field Name"))
```
Comment (dongjoon-hyun) [inline connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala:1078]: ditto

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. I verified Avro module.
```
$ build/sbt "avro/test"
...
[info] Test run org.apache.spark.sql.avro.JavaAvroFunctionsSuite started
[info] Test org.apache.spark.sql.avro.JavaAvroFunctionsSuite.testToAvroFromAvro started
[info] Test run org.apache.spark.sql.avro.JavaAvroFunctionsSuite finished: 0 failed, 0 ignored, 1 total, 0.633s
[info] ScalaTest
[info] Run completed in 2 minutes, 42 seconds.
[info] Total number of tests run: 288
[info] Suites: completed 13, aborted 0
[info] Tests: succeeded 288, failed 0, canceled 0, ignored 2, pending 0
[info] All tests passed.
[info] Passed: Total 289, Failed 0, Errors 0, Passed 289, Ignored 2
[success] Total time: 530 s (08:50), completed Aug 11, 2022 3:04:35 PM
```

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 3.4.0.

**Comment by dongjoon-hyun** [comment]
Thank you, @iemejia , @HyukjinKwon , @xkrogen , @martin-g , @LuciferYang .

**Comment by iemejia** [comment]
Thanks as usual @dongjoon-hyun !

**Pull request closed without merging**